### DISCUSSION TITLE
$env:NODE_OPTIONS="--max-old-space-size=SIZE" adds 48MB for no reason

### DISCUSSION THREAD

**Discussion by Martin-Holos** [nodejs/help#4928] [Q&A]
### Version

17.3.0

### Platform

x64

### What steps will reproduce the bug?

PS C:\> $env:NODE_OPTIONS="--max-old-space-size=8196"
PS C:\> node -e 'console.log(v8.getHeapStatistics().heap_size_limit)'
8644460544

### How often does it reproduce? Is there a required condition?

every time

### What is the expected behavior?

heap-size should be set to 8196MB (8GB)

### What do you see instead?

8644460544 bytes (8244MB)

### Additional information

The LTS version has a similar problem but with 24bytes, which breaks deserialization for 4096

**Comment by nponeccop** [comment]
JFYI 8 GiB = 8192 MiB, not 8196. The observed difference is exactly 48 MiB for some reason though.

8 GiB = 8 * 2^30 = 8192 * 2^20 = 8,589,934,592 bytes
8196 MiB = 8196 * 2^20 = 8,594,128,896 bytes

8644460544 - 8594128896 = 50,331,648 bytes = 48 MiB (exactly)

**Comment by lamweili** [comment]
I have the same issue on NodeJS 14.19.1. 

* `node --max-old-space-size=2048 -e 'console.log(v8.getHeapStatistics().heap_size_limit - (2048 * 1024 * 1024))'`
`50331648`

* `node --max-old-space-size=4096 -e 'console.log(v8.getHeapStatistics().heap_size_limit - (4096 * 1024 * 1024))'`
`50331648`

* `node --max-old-space-size=8192 -e 'console.log(v8.getHeapStatistics().heap_size_limit - (8192 * 1024 * 1024))'`
`50331648`

The difference, 50331648 / 1024 / 1024, is 48 MiB (exactly).

**Comment by bnoordhuis** [comment]
Converted to a discussion because this is working as intended, not a bug.

The garbage collector is a generational collector, i.e., there are two spaces: the young space (a.k.a. the nursery) and the old space. `--max_semi_space_size=..` controls the former, `-max-old-space-size=..` of course the latter.

`--max_semi_space_size=..` defaults to 16 MB but it's tripled because conceptually there is more than one nursery. That's where the extra 48 MB comes from.

**Reply by lamweili** [reply]
Indeed, I traced through and found the same explanation in the codes and tests.

https://github.com/nodejs/node/blob/d65514bc29ff4e91ae560884d10483e33012e095/deps/v8/src/heap/heap.cc#L220-L225

https://github.com/nodejs/node/blob/d65514bc29ff4e91ae560884d10483e33012e095/deps/v8/test/unittests/heap/heap-unittest.cc#L82

**Reply by lamweili** [reply]
Perhaps the documentation needs to be clearer that the [`v8.getHeapStatistics().heap_size_limit`](https://nodejs.org/api/v8.html#v8getheapstatistics) is using the formula above.
Might be closely related to the following documentation improvements as well:

* https://github.com/nodejs/node/issues/39107
* https://github.com/nodejs/node/issues/42511
  * PR https://github.com/nodejs/node/pull/42575

**Reply by bnoordhuis** [reply]
Sure, a pull request would be welcome.
